### ISSUE TITLE
bug(ws1): gpt-oss llama.cpp produces garbled output on Run 2+ of multi-run syllabus (no service restart)

### ISSUE THREAD

**Issue by BittnerPierre** [BittnerPierre/agentic-research#166]
## Real failure mode (corrected)

This is **not** a server crash. The llama.cpp servers (`llm-instruct`, `llm-reasoning`) stay `Up (healthy)` throughout. What fails is **text generation itself**: starting on Run 2 of a multi-run syllabus, the model produces corrupted output — strings of `??...???…???` characters injected into otherwise valid tokens.

The agentic loop then breaks because the corrupted output ends up inside tool-call arguments (e.g. URLs):

\`\`\`json
download_and_store_url_tool({
  \"url\": \"https://medium.com/@callum...??..???…?…..?????...????..??…..???…\"
})
\`\`\`

→ `download_and_store_url` rejects the URL (\"URL invalide (non ASCII)\"), the agent retries, the corruption recurs, and after 10 turns:

\`\`\`
agents.exceptions.MaxTurnsExceeded: Max turns (10) exceeded
  File \"/app/src/deep_research_manager.py\", line 188, in _prepare_knowledge
\`\`\`

So the visible client-side failure is `MaxTurnsExceeded`, but the **root cause** is server-side output degradation.

## Reproduction

Reproduced 2026-04-28 on DGX Spark with the canonical bench command:

\`\`\`bash
./scripts/benchmark-dgx.sh openai \\
  --syllabus test_files/syllabus.md \\
  --runs 3 \\
  --no-report-warmup \\
  --keep-services
\`\`\`

- Run 1: ✅ completes cleanly (237s, score 83.7)
- Run 2: ❌ `MaxTurnsExceeded` in knowledge preparation, garbled tool-call args observed
- Service status post-failure: all `Up (healthy)`

## Setup

- Backend: llama.cpp duo, gpt-oss-20b mxfp4
- Env: `models/models.openai.env` — instruct `--parallel 2 --flash-attn on`, ctx 32k; reasoning `--parallel 1 --flash-attn on --chat-template-kwargs {\"reasoning_effort\":\"medium\"}`, ctx 12k
- Config: `configs/config-docker-dgx.yaml`

## Hypotheses (to discriminate from logs)

1. **KV cache slot reuse with `--parallel 2`** on instruct: a slot from Run 1 is reassigned in Run 2 with stale state, producing degraded sampling.
2. **Slot state contamination** between concurrent slots: two requests sharing `--parallel 2` corrupt each other's logits.
3. **Reasoning channel state** carried over runs (less likely on instruct, but worth checking on reasoning slot too).
4. **Numerical instability with mxfp4** that only surfaces after warmup/long uptime.

Hypothesis 1 or 2 is the leading candidate because (a) it depends on multi-run, (b) `--parallel 2` is unusual, (c) Run 1 is fine.

## Quick diagnostic / mitigation hooks (cheap to test)

- Drop `--parallel 2` → `--parallel 1` on instruct, repeat 3 runs. If green → strong evidence for slot contamination.
- Restart `llm-instruct` between runs (`docker compose restart llm-instruct`) → sanity check that fresh state fixes it.
- Compare with vLLM mono on the same syllabus (separate setup, will be enabled by #169).

## Related

- Issue #144 — different symptom on same target (gpt-oss + syllabus): \"exceeds the available context size\" via litellm. Possibly correlated (long-context degradation), to revisit once root cause here is known.

## Acceptance

- Root cause identified from logs / experiments
- Fix verified by 3 consecutive runs without restart on syllabus.md
- Runbook note + regression check

**Comment by BittnerPierre** [comment]
## Signature affinée (2026-04-28)

Repro discriminée par taille de la charge Run 1.

### Test contrôle 1 — petit report (`query_advanced_1.md`)

\`\`\`bash
./scripts/benchmark-dgx.sh openai --syllabus test_files/query_advanced_1.md --runs 3 --no-report-warmup --keep-services
\`\`\`

✅ **3 runs passent cleanly**, sans restart de service.

- Run avg: 151.1s, score 89.0/100, content 94.8, RAG 84.4
- Pas de génération corrompue

### Test contrôle 2 — gros report (`syllabus.md`)

\`\`\`bash
./scripts/benchmark-dgx.sh openai --syllabus test_files/syllabus.md --runs 3 --no-report-warmup --keep-services
\`\`\`

❌ Run 1 OK (237s, 83.7/100), Run 2 corrompu (URLs garbled → MaxTurnsExceeded en knowledge prep).

### Restart entre runs

✅ Confirmé par l'utilisateur : restart de `llm-instruct` / `llm-reasoning` entre runs corrige systématiquement.

### Conditions nécessaires confirmées

| Condition | Requis ? |
|---|---|
| gpt-oss model | oui (qwen, ministral non affectés) |
| Multi-run sans restart | oui |
| **Run 1 = gros report (syllabus.md)** | oui (query_advanced_1.md ne déclenche pas) |
| Crash dans phase prep de Run 2 | oui (contexte court au moment du crash) |

### Implication mécaniste

Le crash n'est pas dû à la pression contexte instantanée (Run 2 prep = contexte court), mais à un **état rémanent** déposé dans le process `llm-instruct` par un Run 1 suffisamment chargé. Le restart vide cet état → fix.

Hypothèse leader : KV cache fragmenté / slot contamination amplifiée par `--parallel 2`, possiblement combiné avec l'état du reasoning channel propre à gpt-oss (absent de qwen/ministral, ce qui expliquerait la spécificité modèle).

### Prochain test discriminant

`--parallel 2` → `--parallel 1` sur instruct dans `models/models.openai.env`, puis 3× syllabus.md sans restart.

- Si OK → slot contamination confirmée
- Si toujours KO → pister le reasoning channel state ou la mémoire mxfp4

**Comment by BittnerPierre** [comment]
## Upstream research (2026-04-28) — quasi-duplicate found

Search of llama.cpp upstream issues turns up two strong matches; one is essentially the same bug on the same hardware.

### Strong match — same hardware, same models

**ggml-org/llama.cpp#20532 (closed)** — https://github.com/ggml-org/llama.cpp/issues/20532
- DGX Spark + gpt-oss-20b/120b mxfp4 + `--flash-attn` + prompt cache reuse
- First heavy run corrupts subsequent requests on the same llama-server process
- Workaround: `--no-cache-prompt` or `/slots/X?action=erase` between requests
- Closed, but author notes \"any small change to caching or parsing affects manifestation\" → not fully root-cause-fixed

### Root-cause family — gpt-oss + parallel slots + SWA checkpoint

**ggml-org/llama.cpp#17527 (closed)** — https://github.com/ggml-org/llama.cpp/issues/17527
- gpt-oss with `--parallel N`: SWA / hybrid KV checkpoint restore fails after a long first run, requires non-contiguous slot
- ggerganov pointed at the fix (switch `find_slot(ubatch, false)` and adapt reads); not merged at issue close
- Workarounds reported: `--swa-full` (recommended default for gpt-oss anyway), `--kv-unified`

### Other related (weaker)

- #15573 (closed) — gpt-oss-120b multi-turn KV growth + unexpected `<|start|>assistant…` prefix
- #15517 (open) — gpt-oss-120b stalls on follow-up after long answer
- #21523 (open) — generic mangled output, not gpt-oss specific

### Implication for our diagnosis

The previous \"`--parallel 2` slot contamination\" lead was poorly ranked. The actual upstream signature points at **SWA checkpoint + prompt cache reuse on gpt-oss**, exposed by a heavy Run 1, on Spark CUDA hardware specifically. This explains every condition we observed:

- gpt-oss only ✓ (SWA/checkpoint logic specific to gpt-oss arch)
- Heavy Run 1 required ✓ (need to fill cache to expose the bad path)
- Multi-run sans restart ✓ (cache state lives in the server process)
- Restart fixes ✓ (cache wiped)
- mxfp4 not strictly required, but it's what we run

### Revised test plan (cheapest informational test first)

1. **Add `--swa-full` to `LLM_INSTRUCT_EXTRA_PARAMS`** → if 3× syllabus.md passes → confirms #17527 family. Recommended for gpt-oss anyway.
2. If #1 inconclusive: **`--no-cache-prompt`** → if it passes → confirms #20532. Costs perf, used as diagnostic only.
3. Drop `--parallel 2` → `--parallel 1` only if #1 and #2 fail to clarify.

If #1 fixes it cleanly, the resolution is a 1-line env change, no patch to llama.cpp needed.
